File: src/cpp/p2/fifo_barbershop.cpp
#include <mutex>
#include <vector>
#include <thread>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <functional>
#include "cpp/shared/parse.hpp"
#include "cpp/shared/ts_queue.hpp"
#include "cpp/shared/semaphore.hpp"

std::mutex output_mutex;

struct customer_info{
	customer_info(int i = 0, semaphore* s = NULL) : id(i), sem(s) {}
	
	int id;
	semaphore* sem;
};

void customer(int id, ts_queue<customer_info>& queue){
	semaphore sem;
	customer_info info(id, &sem);
	std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 3000));	//Walk to the barbershop...
	
	if(queue.enqueue(info)){	//Shop is not full, enter.
		output_mutex.lock();
		std::cout << "(Customer " << id << ") Enters.\n";	//This isn't perfect, one thread could get the lock first, even though the other go into the queue first.
		output_mutex.unlock();
		
		sem.wait();	//Wait until barber calls you up.
		//Get hair cut...
		sem.wait();	//Wait until barber is done.
	}else{
		output_mutex.lock();
		std::cout << "(Customer " << id << ") The shop is full!\n";		//Shop is full, balk and leave.
		output_mutex.unlock();
	}
}

void barber(ts_queue<customer_info>& queue){
	customer_info next;
	while(queue.dequeue(next)){	//Wait for a customer.
		next.sem->signal();	//Call customer up.
		
		output_mutex.lock();
		std::cout << "(Barber) Customer " << next.id << "!\n";
		output_mutex.unlock();
		std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 3000));	//Cut their hair...
		output_mutex.lock();
		std::cout << "(Barber) All done, customer " << next.id << ".\n";
		output_mutex.unlock();
		
		next.sem->signal();	//Tell customer they're done.
	}
}

void test_scenario(int total_customers, int shop_capacity){
	ts_queue<customer_info> queue(shop_capacity);
	
	std::thread barber_thread(barber, std::ref(queue));
	std::vector<std::thread> customers(total_customers);
	for(int i = 0; i < total_customers; ++i){
		customers.push_back(std::thread(customer, i, std::ref(queue)));
	}
	
	for(auto i = customers.begin(); i != customers.end(); ++i){
		if(i->joinable()){
			i->join();
		}
	}
	
	queue.close();
	if(barber_thread.joinable()){
		barber_thread.join();
	}
}

int main(){
	std::srand(std::time(0));
	try{
		std::cout << "Please input how many customers to run: ";
		int customers = scan_int();
		if(customers >= 0){
			std::cout << "Please input how many chairs there are in the barbershop's waiting room: ";
			int capacity = scan_int();
			if(capacity >= 0){
				test_scenario(customers, capacity);
			}else{
				throw std::invalid_argument("Read a negative value from std::cin.");
			}
		}else{
			throw std::invalid_argument("Read a negative value from std::cin.");
		}
	}catch(const std::invalid_argument& ex){
		std::cout << "Please input a single natural number, and nothing else.\n";
	}
	return 0;
}

File: src/go/p2/fifo_barbershop.go
package main

import (
	"fmt"
	"time"
	"math/rand"
	"go/shared/parse"
)

func customer(id int, queue chan<- int, notify <-chan bool, done chan<- bool) {
	time.Sleep(time.Duration(rand.Intn(10000)) * time.Millisecond)		//Walk to the barbershop...
	
	select {
	case queue <- id:	//Shop is not full, enter.
		fmt.Printf("(Customer %d) Enters.\n", id)
		<- notify	//Wait until barber calls you up.
		//Get hair cut...
		<- notify	//Wait until barber is done.
	default:
		fmt.Printf("(Customer %d) The shop is full!\n", id)		//Shop is full, balk and leave.
	}
	
	done <- true
}

func barber(queue <-chan int, notifiers [](chan bool)) {
	for {
		customer := <- queue			//Wait for a customer.
		notifiers[customer] <- true		//Call customer up.
		
		fmt.Printf("(Barber) Customer %d!\n", customer)
		time.Sleep(time.Duration(rand.Intn(3000)) * time.Millisecond)	//Cut their hair...
		fmt.Printf("(Barber) All done, customer %d.\n", customer)
		
		notifiers[customer] <- true		//Tell customer they're done.
	}
}

func testScenario(totalCustomers, shopCapacity int) {
	queue := make(chan int, shopCapacity)
	notifiers := make([](chan bool), totalCustomers, totalCustomers)
	for i := 0; i < totalCustomers; i++ {
		notifiers[i] = make(chan bool)
	}
	done := make(chan bool)
	
	go barber(queue, notifiers)
	for i := 0; i < totalCustomers; i++ {
		go customer(i, queue, notifiers[i], done)
	}
	
	for i := 0; i < totalCustomers; i++ {
		<- done
	}
}

func main() {
	rand.Seed(time.Now().UTC().UnixNano())
	
	fmt.Print("Please input how many customers to run: ")
	if customers, customersErr := parse.ScanInt(); !customersErr && customers >= 0 {
		fmt.Print("Please input how many chairs there are in the barbershop's waiting room: ")
		if capacity, capacityErr := parse.ScanInt(); !capacityErr && capacity >= 0 {
			testScenario(customers, capacity)
		}else{
			fmt.Println("Please input a single natural number, and nothing else.")
		}
	}else{
		fmt.Println("Please input a single natural number, and nothing else.")
	}
}